rmw_fastrtps_shared_cpp/src/rmw_wait.cpp

```diff
@@ -114,89 +55,94 @@ __rmw_wait(
+        custom_event_info->get_listener()->get_statuscondition());
+      fastdds_wait_set->attach_condition(
+        custom_event_info->get_listener()->event_guard[event->
+        event_type]);
     }
   }
 
   if (guard_conditions) {
     for (size_t i = 0; i < guard_conditions->guard_condition_count; ++i) {
       void * data = guard_conditions->guard_conditions[i];
-      auto guard_condition = static_cast<GuardCondition *>(data);
-      guard_condition->attachCondition(conditionMutex, conditionVariable);
+      auto guard_condition = static_cast<eprosima::fastdds::dds::GuardCondition *>(data);
+      fastdds_wait_set->attach_condition(*guard_condition);
     }
   }
 
-  // This mutex prevents any of the listeners
-  // to change the internal state and notify the condition
-  // between the call to hasData() / hasTriggered() and wait()
-  // otherwise the decision to wait might be incorrect
-  std::unique_lock<std::mutex> lock(*conditionMutex);
-
-  bool hasData = check_wait_set_for_data(
-    subscriptions, guard_conditions, services, clients, events);
-  auto predicate = [subscriptions, guard_conditions, services, clients, events]() {
-      return check_wait_set_for_data(subscriptions, guard_conditions, services, clients, events);
-    };
-
-  bool timeout = false;
-  if (!hasData) {
-    if (!wait_timeout) {
-      conditionVariable->wait(lock, predicate);
-    } else if (wait_timeout->sec > 0 || wait_timeout->nsec > 0) {
-      auto n = std::chrono::duration_cast<std::chrono::nanoseconds>(
-        std::chrono::seconds(wait_timeout->sec));
-      n += std::chrono::nanoseconds(wait_timeout->nsec);
-      timeout = !conditionVariable->wait_for(lock, n, predicate);
-    } else {
-      timeout = true;
-    }
+  eprosima::fastdds::dds::ConditionSeq triggered_coditions;
+  Duration_t timeout{0, 0};
+  if (!no_has_wait) {
+    timeout = (wait_timeout) ?
+      Duration_t{static_cast<int32_t>(wait_timeout->sec),
+      static_cast<uint32_t>(wait_timeout->nsec)} : eprosima::fastrtps::c_TimeInfinite;
   }
 
-  // Unlock the condition variable mutex to prevent deadlocks that can occur if
-  // a listener triggers while the condition variable is being detached.
-  // Listeners will no longer be prevented from changing their internal state,
-  // but that should not cause issues (if a listener has data / has triggered
-  // after we check, it will be caught on the next call to this function).
-  lock.unlock();
+  ReturnCode_t ret_code = fastdds_wait_set->wait(
+    triggered_coditions,
+    timeout
+  );
 
   if (subscriptions) {
     for (size_t i = 0; i < subscriptions->subscriber_count; ++i) {
       void * data = subscriptions->subscribers[i];
       auto custom_subscriber_info = static_cast<CustomSubscriberInfo *>(data);
-      custom_subscriber_info->listener_->detachCondition();
-      if (!custom_subscriber_info->listener_->hasData()) {
+      eprosima::fastdds::dds::StatusCondition & status_condition =
```

Comment:
I think we don't need to check if the condition has been triggered, just if there is a sample ready to be taken.

Applies to services and clients as well.

Reply:
Done in 5ff6c5e2